fix!(copilot): data parsing when stream is false

lucobellic opened the pull request:
## Description

<!-- Describe the big picture of your changes to communicate to the maintainers
  why we should accept this pull request. -->

Update openai adapter data parsing (also used by copilot) when stream is set to false.

When stream is set to false the format do not start with `data: ` and message content is in `json.choices[1].message.content` (see #261 log output for the format)

## Related Issue(s)

- Fixes #261

## Screenshots

<!-- Add screenshots of the changes if applicable. -->

## Checklist

- [x] I've read the [CONTRIBUTING](https://github.com/olimorris/codecompanion.nvim/blob/main/CONTRIBUTING.md) guidelines.

olimorris commented:
Thanks for this.

Out of curiosity what's your use case for not streaming? And wouldn't it be better to change the method in `http.lua`?

lucobellic commented:
> Thanks for this.
> 
> Out of curiosity what's your use case for not streaming? Would it make sense to also check for the `stream = false` parameter?

I had to disable stream (among other things) in order to use o1 models (o1-preview and o1-mini) with copilot.
Otherwise `Bad Request` is always returned.

I will had a check on the stream parameter instead of using `data` search and fallback logic.

S1M0N38 commented:
> > Thanks for this.
> > Out of curiosity what's your use case for not streaming? Would it make sense to also check for the `stream = false` parameter?
> 
> I had to disable stream (among other things) in order to use o1 models (o1-preview and o1-mini) with copilot. Otherwise `Bad Request` is always returned.
> 
> I will had a check on the stream parameter instead of using `data` search and fallback logic.

Are o1 family already available in copilot?

lucobellic commented:
@S1M0N38 o1 models (preview, mini) are available with Copilot (with limited usage) if you sign up for early access (on a waiting list).

lucobellic commented:
@olimorris, to add a check for `stream = false`, I need to pass `self` to `chat_output` and `inline_output`. Should I proceed with this change, or were you thinking of something else?
Also feel free to directly edit the PR.

olimorris commented:
> @olimorris, to add a check for `stream = false`, I need to pass `self` to `chat_output` and `inline_output`. Should I proceed with this change, or were you thinking of something else?
> 
> Also feel free to directly edit the PR.

Yeah I think that makes the most sense. If I recall (I'm away right now and using my phone) we should be able to pass `self.adapter` to both with minimal fuss.

lucobellic commented:
Updates:
 - Added self parameter to all handlers function `inline_output`, `chat_output` and `tokens`.
 - Updated documentation
 - Updated logic to extract data based on `self.parameters.stream`
 
> [!NOTE]
> Latest changes introduce a breaking change

olimorris commented:
Firstly, thank you so much for your diligence with this PR. It all works perfectly having tested it out.

Where do you think the breaking change is btw? Just in the way that adapters now work and user's will have to accommodate for `self`?

lucobellic commented:
> Where do you think the breaking change is btw? Just in the way that adapters now work and user's will have to accommodate for `self`?

Yes, exactly. I see the breaking changes in the added `self` parameter since these functions are part of the public API and user-extended/custom adapter handlers might not work anymore.

Otherwise, I can remove the `!` from commits/PR if you don't see this as a breaking change.

olimorris commented:
Thanks again for this! Really appreciated
